File: circular-buffer/circular_buffer.rb
class CircularBuffer

  class BufferEmptyException < Exception
  end

  class BufferFullException < Exception
  end

  def initialize(size)
    @size = size
    init_buffer
  end

  def read
    raise BufferEmptyException if buffer.compact.empty?
    result = buffer.at(read_index)
    buffer[read_index] = nil
    advance_read_index
    result
  end

  def write(val)
    raise BufferFullException if full?
    buffer[write_index] = val
    advance_write_index
  end

  def write!(val)
    prev_full = full?
    buffer[write_index] = val
    advance_write_index
    advance_read_index if prev_full
  end

  def clear
    init_buffer
  end

  private

  attr_accessor :buffer, :read_index, :write_index, :size

  def init_buffer
    @buffer = Array.new(size)
    @read_index = 0
    @write_index = 0
  end

  def advance_read_index
    self.read_index = (read_index + 1) % size
  end

  def advance_write_index
    self.write_index = (write_index + 1) % size
  end

  def full?
    buffer.compact.length == size
  end

end

File: go/circular-buffer/circular_buffer.go
package circular

import "errors"

type Buffer struct {
  buf        []byte
  size       int
  readIndex  int
  writeIndex int
}

func (b *Buffer) advanceReadIndex() {
  b.readIndex = (b.readIndex + 1) % b.size
}

func (b *Buffer) advanceWriteIndex() {
  b.writeIndex = (b.writeIndex + 1) % b.size
}

func (b *Buffer) ReadByte() (byte, error) {
  result := b.buf[b.readIndex]
  if result == 0 {
    return 0, errors.New("Nothing to read")
  }

  b.buf[b.readIndex] = 0
  b.advanceReadIndex()

  return result, nil
}

func (b *Buffer) WriteByte(c byte) error {
  if b.buf[b.writeIndex] != 0 {
    return errors.New("Buffer is full")
  }

  b.buf[b.writeIndex] = c
  b.advanceWriteIndex()

  return nil
}

func (b *Buffer) Overwrite(c byte) {
  full := b.buf[b.writeIndex] != 0

  b.buf[b.writeIndex] = c
  b.advanceWriteIndex()
  if full {
    b.advanceReadIndex()
  }

  return
}

func (b *Buffer) Reset() {
  b.buf = make([]byte, b.size)
  b.readIndex = 0
  b.writeIndex = 0
}

func NewBuffer(size int) *Buffer {
  return &Buffer{buf: make([]byte, size), size: size}
}
